### PR TITLE
crmRouteBinder - Remove params from url if they equal their defaults

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Reduces url clutter in the Api4 explorer and fixes a bug with `null` being cast to `0`.

Before
----------------------------------------
When removing a param in the api4 explorer, it persists in the url. This is a problem in the case of the `Relationship` api and the `$current` param. To reproduce:

1. Go to the api explorer and select `Relationship` `get`.
2. Click a `Current` radio button.
3. Click the `x` to clear the `Current` radio selection.
4. Note that `current=0` is still in the url.
5. Refresh the page and the `false` radio will be selected inappropriately.

After
----------------------------------------
Clearing a selection from the api explorer will remove it from the url, which reduces clutter and prevents errors.

### CODE REVIEW

File: ang/crmRouteBinder.js [hunk 1]
```diff
@@ -50,18 +50,18 @@
         registerGlobalListener($injector);
 
         options.format = options.format || 'json';
-        var fmt = formats[options.format];
+        var fmt = _.clone(formats[options.format]);
         if (options.deep) {
           fmt.watcher = '$watch';
```
Comment (colemanw) [inline ang/crmRouteBinder.js]: Adding `_.clone()` on l53 prevents line 55 here from modifying the global defaults.

File: ang/crmRouteBinder.js [hunk 2]
```diff
@@ -50,18 +50,18 @@
...
         }
-        var _scope = this;
+        var value,
+          _scope = this,
+          $route = $injector.get('$route'),
+          $timeout = $injector.get('$timeout');
 
-        var $route = $injector.get('$route'), $timeout = $injector.get('$timeout');
-
-        var value;
```
Comment (colemanw) [inline ang/crmRouteBinder.js]: Just some cleanup

File: ang/crmRouteBinder.js [hunk 3]
```diff
@@ -74,12 +74,25 @@
...
           pendingUpdates = pendingUpdates || {};
           pendingUpdates[options.param] = encValue;
           var p = angular.extend({}, $route.current.params, pendingUpdates);
-          angular.forEach(ignorable, function(v,k){ if (p[k] === v) delete p[k]; });
+
+          angular.forEach(ignorable, function(v, k) {
+            if (p[k] === v) {
+              delete p[k];
+            }
+          });
```
Comment (colemanw) [inline ang/crmRouteBinder.js]: Just cleanup - expanded the 1-liner on l82 to be more readable.

**Comment by totten** [comment]
I haven't tested, but want to give a conceptual :+1: -- if the same value would be loaded as a default, then it's preferable to appear shorter and normalized.

For someone `r-run`ning this, it's worth considering the different paths that can bring one to a route, e.g.:

* Clicking a widget in a GUI that updates a filter and automatically updates the URL
* Clicking backward and forward buttons
* Opening the base-page, clicking around, and then editing the URL  to specify a different set of filters/route.
* Opening the base-page and clicking a hyperlink which goes to a different set of filters/route.
* Starting from an external page -- and following a hyperlink which has a bunch of parameters

**Pull request merged**